clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java

```diff
@@ -0,0 +1,127 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.clients;
+
+import org.apache.kafka.common.errors.UnsupportedVersionException;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.Protocol;
+import org.apache.kafka.common.requests.ApiVersionsResponse;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.TreeMap;
+
+public class NodeApiVersions {
+    private final Collection<ApiVersionsResponse.ApiVersion> apiVersions;
+
+    private final static ApiKeys[] ID_TO_APIKEY;
+
+    static {
+        ID_TO_APIKEY= new ApiKeys[ApiKeys.MAX_API_KEY + 1];
+        for (ApiKeys key : ApiKeys.values())
+            ID_TO_APIKEY[key.id] = key;
+    }
+
+    // An array of the usable versions of each API, indexed by ApiKeys ID.
+    private final short[] usableVersions;
+
+    public NodeApiVersions(Collection<ApiVersionsResponse.ApiVersion> apiVersions) {
+        this.apiVersions = apiVersions;
+        this.usableVersions = new short[ID_TO_APIKEY.length];
+        Arrays.fill(usableVersions, (short) -1);
+        for (ApiVersionsResponse.ApiVersion apiVersion: apiVersions) {
+            int index = apiVersion.apiKey;
+            if ((index < 0) || (index >= usableVersions.length)) {
+                continue;
+            }
+            if (Protocol.CURR_VERSION[index] < apiVersion.minVersion) {
+                usableVersions[index] = -1;
+                continue;
+            }
+            if (Protocol.MIN_VERSIONS[index] > apiVersion.maxVersion) {
+                usableVersions[index] = -1;
+                continue;
+            }
+            if (Protocol.CURR_VERSION[index] < apiVersion.maxVersion) {
+                usableVersions[index] = Protocol.CURR_VERSION[index];
+            } else {
+                usableVersions[index] = apiVersion.maxVersion;
+            }
+        }
+    }
+
+    public short getUsableVersion(short apiKey) {
```

Comment:
Could we have this accept an `ApiKeys` instance instead of short? Then we wouldn't need the check for a negative api key. Also would be nice to document what exactly this method is returning. Is it the largest version supported by both the client and server?